server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java

```diff
@@ -570,6 +572,14 @@ public boolean delete(TemplateProfile profile) {
         if (success) {
             if ((imageStores != null && imageStores.size() > 1) && (profile.getZoneIdList() != null)) {
                 //if template is stored in more than one image stores, and the zone id is not null, then don't delete other templates.
+                List<VMTemplateZoneVO> templateZones = templateZoneDao.listByTemplateId(template.getId());
+                List<Long> zoneIds = templateZones.stream().map(VMTemplateZoneVO::getZoneId).collect(Collectors.toList());
+                if (zoneIds.size() > 0) {
+                    return success;
+                }
+                template.setRemoved(new Date());
+                template.setState(State.Inactive);
+                templateDao.update(template.getId(), template);
```

Comment:
`delete()` ranges from line 454 to line 621. It deserves its own worker class. I won't ask that (hint hint, wink wink) but can you at least extract your new bit to a separate method?

Reply:
done @DaanHoogland